### ISSUE TITLE
case study: how to index buildings with name

### ISSUE THREAD

**Issue by rinigus** [Framstag/libosmscout#238]
I think that I am missing something trivial. Maybe you could point me in the right direction. I am missing a building that has a name in OSM, in addition to address. It seems that only address is imported into Location index, but not the name.

See http://www.openstreetmap.org/way/104797987

User would expect to be able to search by "Ahhaa keskus", not "Sadama 1".

**Comment by Karry** [comment]
I think that it is a workaround, not solution for problem. Buildings are not POIs from my point of view, even if it has has name. Shopping centers or tourism attractions are...

I think that everything with name should be added to `textother.dat` and should be available in search... It seems that it dont work like this currently, I cant find ponds with name...

**Comment by Karry** [comment]
The linked building "Ahhaa keskus" is shopping center? Conference center or something other? If it should be POI, there should be another tag in OSM that define type of POI.

**Comment by Framstag** [comment]
The information is there:

```
[tim:~/projects/OSMScout/DumpData] src/DumpData ../maps/estonia -w 104797987
Area {
  OSM id: 104797987
  fileOffset: 25434096
  type: building
  boundingBox: [58,37615 N 26,73386 E - 58,37694 N 26,73528 E]
  center: 58,37654 N 26,73457 E
  cell level: 17

  Name: Ahhaa keskus
  Address: 1

  role[0] {
    outer
    type: building
    boundingBox: [58,37615 N 26,73386 E - 58,37694 N 26,73528 E]
    center: 58,37654 N 26,73457 E
    Name: Ahhaa keskus
    Address: 1

    node[0] { lat: 58.37694 lon: 26.73440 }
    node[1] { lat: 58.37680 lon: 26.73480 }
    node[2] { lat: 58.37660 lon: 26.73516 }
    node[3] { lat: 58.37653 lon: 26.73504 }
    node[4] { lat: 58.37643 lon: 26.73510 }
    node[5] { lat: 58.37643 lon: 26.73516 }
    node[6] { lat: 58.37624 lon: 26.73528 }
    node[7] { lat: 58.37615 lon: 26.73466 }
    node[8] { lat: 58.37615 lon: 26.73454 }
    node[9] { lat: 58.37653 lon: 26.73386 }
    node[10] { lat: 58.37659 lon: 26.73396 }
  }
}
```

Street information is only available in the LocationIndex.

The style sheet tells the renderer to draw both:

```
...
     [FEATURE Address] {
       NODE.TEXT#address { label: Address.name; color: @buildingLabelColor; size: 0.7; priority: @labelPrioBuilding; position: 1; }
       AREA.TEXT#address { label: Address.name; color: @buildingLabelColor; size: 0.7; priority: @labelPrioBuilding; position: 1; }
     }

     [TYPE building] {
       AREA.TEXT#buildingName { label: Name.name; color: @buildingLabelColor; size: 0.7; priority: @labelPrioBuilding; }
     }
...
```

And the name and house number is rendered.

If you look at LocationIndexGenerator::IndexAddressAreas() you see that areas are added to the index, if..
* They have an address feature
* They have a name and are a POI

What is the definition of POI? In the context of libosmscout a POI is something the creator of the map.ost finds important and that should be searchable by it's name. Most of the time it is so important that people know only its name but not their address. Thing of stadium, railway station, import building (eifeltower, museum,...).

It is up to you to define "just "buildings with names" as POI. IMHO not every building with a name is important. But you may think different. And it is the advantage of libosmscout that we do not need to agree :-) Also I think in this case the mapping is not optimal. If it *is* important, it should have a better tagging, using something like shop_XXX, sport_XXX, historic_XXX, tourism_XXX, amenity_XXX. This would tag it obviously as "somehow important". And exactly all this types have "POI" in the map.ost. If we tag everything with a name as POI, than we would not need the POI flag in the map.ost at all?

**Comment by rinigus** [comment]
Just for background: since I use libosmscout as an importer to feed libpostal-based geocoder, my requirements are not exactly the same as usual. So, I am looking for the similar data coverage (or more) that includes Location and free text search (MARISA). By running all through Location, I get access to admin hierarchy for all searches (including free text search) which is a great improvement for me. Right now I am regenerating world-covering databases that would fix few early issues (slow search in Germany for common names and some missing POI data, including adding POI to all buildings).

Let me address all the questions:

"Ahhaa keskus": its a science center / museum. Probably similar issue would be with some shopping centers. As you could see, they don't have any additional tags, just name.

Whether I think its POI or not: I agree that the flexibility of libosmscout is great and its actually a good example where it can be put into great use. I think that almost anything with the name should be POI - how do I know what the user would search for? In some places, addresses of the houses are through the given names. For this reason, I enabled name search for all buildings that have it defined (POI in ost). 

When we talk about enabling all names as POIs in libosmscout, it may hinder the location search performance too much. Probably, it needs to be tested and maybe has been tested already?

**Comment by Framstag** [comment]
I did not test the location index for "index all" scenarios. I would assume that the performance is more linear than better. The location index within a region is a simple list and for some lookups sub trees are traversed. So no "n log n" but worse. The question though is "how much will the actual index increase". I assume that there not that many builds with names that are not already index? So if you are lucky, there only 10% increase in size.

**Comment by rinigus** [comment]
I think in the case of Estonia, the increase was marginal, but margins could have been +/-10%. For a planet, its hard to say - I did not record the sizes of the imports. However, when I allowed to index all objects with a name, the location index that I have in libpostal-based geocoder increased by 45% of the original size. Assuming that the geocoder database increases linearly with the objects, I could see that having all objects as POIs maybe not as desirable in all cases.
